modin/core/execution/ray/generic/partitioning/partition_manager.py

```diff
@@ -40,15 +40,19 @@ def to_numpy(cls, partitions, **kwargs):
         -------
         NumPy array
         """
-        parts = RayWrapper.materialize(
-            [
-                obj.apply(lambda df, **kwargs: df.to_numpy(**kwargs)).list_of_blocks[0]
-                for row in partitions
-                for obj in row
-            ]
-        )
-        n = partitions.shape[1]
-        parts = [parts[i * n : (i + 1) * n] for i in list(range(partitions.shape[0]))]
-
-        arr = np.block(parts)
-        return arr
+        if partitions.shape[1] == 1:
```

Comment:
are you sure that it's only beneficial for frames having only one column partition? I tried running your changes on data with the shape of (1_000_000, 36) that has 2 col partitions and got the following results: 
![image](https://github.com/modin-project/modin/assets/62142979/84485458-4341-44e5-9b47-45282f539b55)
I also modified the reproducer a bit, don't know whether it affected the results:
<details><summary>code</summary>

```python
import modin.pandas as pd
import numpy as np
from time import time
from modin.utils import execute

import modin.config as cfg

# init the workers
pd.DataFrame([cfg.MinPartitionSize.get() * cfg.NPartitions.get()]).to_numpy()

for _ in range(5):
    df = pd.DataFrame(np.random.rand(10 ** 6, 36).astype(str))
    execute(df)
    start = time()
    arr = df.to_numpy()
    print(f"to_numpy conversion: {time()-start}, {arr.size=}")
```

</details>

could you please run more tests with different data sizes/partition numbers/data types?

Reply:
Ray (112 cpus):

<img width="272" alt="image" src="https://github.com/modin-project/modin/assets/45976948/391b81b4-ac21-4919-ad98-d9722b1909e0">

<details>

<summary> code </summary>

```python
import warnings
import modin.pandas as pd
import numpy as np
from time import time
from modin.utils import execute

import modin.config as cfg
warnings.simplefilter("ignore", category=UserWarning)

# init the workers
pd.DataFrame([cfg.MinPartitionSize.get() * cfg.NPartitions.get()]).to_numpy()

sizes = [
    (10 ** 6, 10), (10 ** 6, 10 ** 2),
    (10 ** 7, 10), (10 ** 7, 10 ** 2),
    (10 ** 3, 10 ** 3), (10 ** 4, 10 ** 4),
]

types = [str, object]

for size in sizes:
    for _type in types:
        iterations = 5
        times = [None] * iterations
        start = time()
        df = pd.DataFrame(np.random.rand(*size)).astype(_type)
        execute(df)
        print(f"DataFrame created: {round(time()-start, 2)}", end=" |")
        for idx in range(iterations):
            start = time()
            arr = df.to_numpy()
            times[idx] = round(time()-start, 3)
        print(f"{size=}, {_type=}: {times=}, {min(times)=}")

```

</details>

Reply:
ok, let's leave as it is for now, but I would say that the changes have potential to speed-up more cases if we investigate more

Reply:
Agree, although this will require knowledge of types at this level, or making a logic dispatcher at the dataframe level.